File: solutions/53/02.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const maxSubArray = nums => {
  let currSum = -Infinity, maxSum = currSum;

  nums.forEach(num => {
    currSum = Math.max(num, currSum + num);
    maxSum = Math.max(maxSum, currSum);
  });

  return maxSum;
};

const testSetup = [
  [[-2, 1, -3, 4, -1, 2, 1, -5, 4], 6],
  [[2], 2],
  [[-5, -5, 2], 2],
  [[2, -8, 3, -2, 4, -10], 5]
], eq = require('assert').strictEqual;

for(const [args, ans] of testSetup)
  eq(maxSubArray(args), ans);

File: solutions/53/02.py
import math

def f(A):
    curr_suba_sum = -math.inf
    max_sum = -math.inf

    for e in A:
        curr_suba_sum = max(e, curr_suba_sum + e)
        max_sum = max(max_sum, curr_suba_sum)

    return max_sum

assert f([2]) == 2
assert f([-2, 1, -3, 4, -1, 2, 1, -5, 4]) == 6
assert f([-5, -5, 2]) == 2
